File: src/leetcode/trees/preorder/SymmetricTree.java
package leetcode.trees.preorder;

import leetcode.trees.TreeNode;

public class SymmetricTree {
    public boolean isSymmetric(TreeNode root) {
        if (root == null) {
            return true;
        }
        return isSymmetricSubtree(root.left, root.right);
    }

    private boolean isSymmetricSubtree(TreeNode left, TreeNode right) {
        if (left == null && right == null) {
            return true;
        }
        if (left == null || right == null) {
            return false;
        }
        return left.val == right.val && isSymmetricSubtree(left.left, right.right) && isSymmetricSubtree(left.right, right.left);
    }

}

File: src/leetcode/trees/preorder/SymmetricTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isSymmetric = function(root) {
    if (!root) {
        return true;
    }
    return isSymmetricSub(root.left, root.right);
};

const isSymmetricSub = (left, right) => {
    if (!left && !right) {
        return true;
    }
    if (!left || !right) {
        return false;
    }
    return left.val === right.val && isSymmetricSub(left.left, right.right) && isSymmetricSub(left.right, right.left);
};